### ISSUE TITLE
When using scrcpy 3.0 on some devices, no new SPS frames are generated when the device is rotated.

### ISSUE THREAD

**Issue by sbfkcel** [Genymobile/scrcpy#5537]
## Environment

 - **OS:** macOS
 - **Scrcpy version:** 3.0
 - **Installation method:** github release
 - **Device model:** HUAWEI ANA-AN00
 - **Android version:** 12

## Describe the bug

When the device screen rotates, the scrcpy screen does not rotate.
If you use scrcpy 2.7, everything is normal.

After preliminary investigation, when using scrcpy 3.0, when the device screen rotates, the SPS frame should change, but no new SPS frame is generated on this device.

```
➜ scrcpy
scrcpy 3.0 <https://github.com/Genymobile/scrcpy>
INFO: ADB device found:
INFO:     -->   (usb)  KXU0221408005112                device  ANA_AN00
/opt/homebrew/Cellar/scrcpy/3.0/share/scrcpy/scrcpy-server: 1 file pushed, 0 skipped. 133.6 MB/s (90224 bytes in 0.001s)
[server] INFO: Device: [HUAWEI] HUAWEI ANA-AN00 (Android 12)
INFO: Renderer: metal
INFO: Texture: 1080x2336
2024-11-26 09:48:43.848 scrcpy[30447:5761806] +[IMKClient subclass]: chose IMKClient_Modern
2024-11-26 09:48:43.848 scrcpy[30447:5761806] +[IMKInputSession subclass]: chose IMKInputSession_Modern
[server] ERROR: Exception on thread Thread[DisplayListener,5,main]
java.lang.NullPointerException: Attempt to invoke virtual method 'boolean java.lang.String.contains(java.lang.CharSequence)' on a null object reference
	at android.hardware.display.DisplayManagerGlobal$DisplayListenerDelegate.handleMessage(DisplayManagerGlobal.java:1000)
	at android.os.Handler.dispatchMessage(Handler.java:117)
	at android.os.Looper.loopOnce(Looper.java:205)
	at android.os.Looper.loop(Looper.java:293)
	at android.os.HandlerThread.run(HandlerThread.java:110)
```
<img width="708" alt="image" src="https://github.com/user-attachments/assets/8105b6a2-565b-4cb3-9093-02ac631cefb7">

<img width="335" alt="image" src="https://github.com/user-attachments/assets/ec129566-ff4f-4d26-a860-9b648d7bcafc">

**Comment by rom1v** [comment]
Please post your framework.jar:

```
adb pull /system/framework/framework.jar
```

**Comment by sbfkcel** [comment]
[Uploading framework.jar.zip…]()

**Comment by rom1v** [comment]
The custom ROM on your device added:
```java
                        if (this.mListener.toString().contains(DisplayManagerGlobal.TARGET_PACKAGE_NAME)) {
                            Log.i(DisplayManagerGlobal.TAG, "Prepare Notification launcher onDisplayChanged");
                        }
```

This should fix the issue:

```java
diff --git server/src/main/java/com/genymobile/scrcpy/wrappers/DisplayManager.java server/src/main/java/com/genymobile/scrcpy/wrappers/DisplayManager.java
index b497e97fa..d44ac608f 100644
--- server/src/main/java/com/genymobile/scrcpy/wrappers/DisplayManager.java
+++ server/src/main/java/com/genymobile/scrcpy/wrappers/DisplayManager.java
@@ -192,6 +192,9 @@ public final class DisplayManager {
                         if ("onDisplayChanged".equals(method.getName())) {
                             listener.onDisplayChanged((int) args[0]);
                         }
+                        if ("toString".equals(method.getName())) {
+                            return "DisplayListener";
+                        }
                         return null;
                     });
             try {
```

Please test by replacing this file in scrcpy 3.0:

- [`scrcpy-server`](https://tmp.rom1v.com/scrcpy/5537/1/scrcpy-server) <sub>`SHA-256: 5c9c99c4c8efa6dddc887a0f92c62837ec0cceaf398475689de272cc4954c8a`</sub>

**Comment by sbfkcel** [comment]
Cool, tested and verified that the issue is fixed.

**Comment by rom1v** [comment]
Merged: 3e689020baa1b3ea1b66cba3260a7a33be458a06
